worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/segment/SegmentMapPartitionDataReader.java

```diff
@@ -0,0 +1,324 @@
+    synchronized (lock) {
+      if (isReleased || buffersToSend.isEmpty()) {
+        return null;
+      }
+
+      RecyclableBuffer buffer = null;
+      int numCredit = credits.get();
+      if (numCredit > 0) {
+        buffer = buffersToSend.poll();
+      }
+      if (numCredit <= 1) {
+        notifyBacklog(getBacklog());
+      }
+      return buffer;
+    }
+  }
+
+  @Override
+  protected void notifyBacklog(int backlog) {
+    if (backlog == 0) {
+      return;
+    }
+    super.notifyBacklog(backlog);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("SegmentMapDataPartitionReader{");
+    sb.append("startPartitionIndex=").append(startPartitionIndex);
+    sb.append(", endPartitionIndex=").append(endPartitionIndex);
+    sb.append(", streamId=").append(streamId);
+    sb.append('}');
+    return sb.toString();
+  }
+
+  private boolean isEndOfSegment(ByteBuf buffer, int subPartitionId) {
+    boolean isEndOfSegment = false;
+    Preconditions.checkState(
+        subPartitionId >= startPartitionIndex && subPartitionId <= endPartitionIndex);
+    if (mapFileMeta.hasPartitionSegmentIds()) {
+      // skip another 3 int fields, the write details are in
+      // FlinkShuffleClientImpl#pushDataToLocation
+      buffer.skipBytes(3 * 4);
+      int dataType = buffer.readByte();
+      isEndOfSegment = END_OF_SEGMENT_BUFFER_DATA_TYPE == dataType;
+    }
+    return isEndOfSegment;
+  }
+
+  private int getBacklog() {
+    synchronized (lock) {
+      Integer backlog = backlogs.peekFirst();
+      while (backlog != null && backlog == 0) {
+        backlogs.pollFirst();
+        backlog = backlogs.peekFirst();
+      }
+      if (backlog != null) {
+        backlogs.pollFirst();
+      }
+      return backlog == null ? 0 : backlog;
+    }
+  }
+
+  @GuardedBy("lock")
+  private void addNewBacklog() {
+    backlogs.addLast(0);
+  }
+
+  @GuardedBy("lock")
+  private void increaseBacklog() {
+    Integer backlog = backlogs.pollLast();
+    if (backlog == null) {
+      backlogs.addLast(1);
+    } else {
+      backlogs.addLast(backlog + 1);
+    }
+  }
+
+  private void updateSegmentId(int subPartitionId) {
+    synchronized (lock) {
+      // Note that only when writing buffers, it has the segment info, when loading buffers from
+      // disk,
+      // we do not know when the segment is started, so we try to get the segment id from the buffer
+      // index here.
+      Integer segmentId =
+          mapFileMeta.getSegmentIdByFirstBufferIndex(
+              subPartitionId, subPartitionNextBufferIndex.get(subPartitionId));
+      if (segmentId == null) {
+        return;
+      }
+      if (segmentId != -1) {
+        // For the continuous segments, we use the same backlog
+        if (segmentId == 0
+            || !segmentId.equals(subPartitionLastSegmentIds.get(subPartitionId))
+                && segmentId != (subPartitionLastSegmentIds.get(subPartitionId) + 1)) {
```

Comment:
Could `subPartitionLastSegmentIds.get(subPartitionId)` return null?

Reply:
No, we already set the default value(-1) for every-subpartition.